include/seastar/util/log-impl.hh

```diff
@@ -50,40 +50,43 @@ private:
     void realloc_buffer();
 
 public:
+    // inserter_iterator is designed like std::back_insert_iterator:
+    // operator*, operator++ and operator++(int) are no-ops,
+    // and all the work happens in operator=, which pushes a character
+    // to the buffer.
+    // The iterator stores no state of its own.
+    //
+    // inserter_iterator is supposed to be used as an output_iterator.
+    // That is, assignment is expected to alternate with incrementing.
     class inserter_iterator {
     public:
         using iterator_category = std::output_iterator_tag;
         using difference_type = std::ptrdiff_t;
-        using value_type = char;
-        using pointer = char*;
-        using reference = char&;
+        using value_type = void;
+        using pointer = void;
+        using reference = void;
 
     private:
         log_buf* _buf;
-        char* _current;
 
     public:
-        explicit inserter_iterator(log_buf& buf) noexcept : _buf(&buf), _current(_buf->_current) { }
-        inserter_iterator(const inserter_iterator& o) noexcept : _buf(o._buf), _current(o._current) { }
+        explicit inserter_iterator(log_buf& buf) noexcept : _buf(&buf) { }
 
-        reference operator*() {
-            if (__builtin_expect(_current == _buf->_end, false)) {
+        inserter_iterator& operator=(char c) {
+            if (__builtin_expect(_buf->_current == _buf->_end, false)) {
                 _buf->realloc_buffer();
-                _current = _buf->_current;
             }
-            return *_current;
+            *_buf->_current++ = c;
+            return *this;
+        }
+        inserter_iterator& operator*() {
+            return *this;
         }
         inserter_iterator& operator++() noexcept {
```

Comment:
I think we have a problem here, in the pre-increment operator.
Why do we allow it at all?
If the user does
  *(++i) = 3;
This will *NOT* do the right thing. There is not even a right thing to do here (what does it mean to increment the iterator without pushing anything to the list?). So wouldn't it be better that this fails to compile?

Reply:
> Here, I don't understand what's the point of defining a type like "reference" if we define it to the "wrong" or "not supposed to be used" type. What forces us to define these aliases at all?

https://en.cppreference.com/w/cpp/named_req/Iterator#Requirements